In [1]:
pip install pyotp

In [2]:
import hashlib

# Mật khẩu gốc (đã băm trước đó) - ví dụ: "mypassword"
stored_password = hashlib.sha256(b"mypassword").hexdigest()

# Nhập mật khẩu từ người dùng
password = input("Nhập mật khẩu: ")

# Băm mật khẩu người dùng nhập
hashed_password = hashlib.sha256(password.encode()).hexdigest()

# So sánh mật khẩu băm với mật khẩu đã lưu
if hashed_password == stored_password:
    print(" Xác thực thành công!")
else:
    print(" Xác thực thất bại!")


Nhập mật khẩu: 12345678910
 Xác thực thất bại!


In [5]:
import pyotp
import time

# Tạo khóa bí mật và mã OTP
secret = pyotp.random_base32()
totp = pyotp.TOTP(secret)

print("Khóa bí mật (dùng để cài ứng dụng Google Authenticator):", secret)
print("Mã OTP của bạn là:", totp.now())

# Yêu cầu người dùng nhập mã OTP
otp_input = input("Nhập mã OTP: ")

# Xác thực mã OTP
if totp.verify(otp_input):
    print("Xác thực thành công!")
else:
    print(" Xác thực thất bại!")


Khóa bí mật (dùng để cài ứng dụng Google Authenticator): QSIO3PEP37SFMOJ5SMNHCRVVM3F5224Q
Mã OTP của bạn là: 161275
Nhập mã OTP: 161275
Xác thực thành công!


In [6]:
import hashlib
import pyotp
import time

# Bước 1: Xác thực bằng mật khẩu
stored_password = hashlib.sha256(b"mypassword").hexdigest()  # Mật khẩu lưu dưới dạng SHA-256 hash

password = input("Nhập mật khẩu: ")
hashed_password = hashlib.sha256(password.encode()).hexdigest()

if hashed_password == stored_password:
    print(" Xác thực mật khẩu thành công! Chuyển sang bước xác thực bằng mã OTP.")
else:
    print(" Xác thực mật khẩu thất bại!")
    exit()  # Thoát nếu sai mật khẩu

# Bước 2: Xác thực bằng mã OTP
# Tạo khóa bí mật
secret = pyotp.random_base32()
totp = pyotp.TOTP(secret)

# In mã OTP (trong thực tế sẽ gửi qua SMS hoặc email)
print("Mã OTP của bạn là:", totp.now())

# Yêu cầu người dùng nhập mã OTP
otp_input = input("Nhập mã OTP: ")

# Xác thực OTP
if totp.verify(otp_input):
    print(" Xác thực hai yếu tố thành công!")
else:
    print(" Xác thực bước 2 (OTP) thất bại!")


Nhập mật khẩu: 123
 Xác thực mật khẩu thất bại!
Mã OTP của bạn là: 499273
Nhập mã OTP: 499273
 Xác thực hai yếu tố thành công!


 1.  Tại sao xác thực hai yếu tố (2FA) lại an toàn hơn so với xác thực chỉ bằng mật khẩu?

 Xác thực hai yếu tố (2FA) an toàn hơn vì nó yêu cầu người dùng phải cung cấp hai loại thông tin khác nhau: thứ bạn biết (mật khẩu) và thứ bạn có (mã OTP từ thiết bị cá nhân). Nếu chỉ dùng mật khẩu, khi bị rò rỉ hoặc đoán đúng, tài khoản sẽ bị chiếm đoạt. Tuy nhiên, với 2FA, ngay cả khi kẻ xấu biết mật khẩu thì họ vẫn cần có thiết bị chứa mã OTP (như điện thoại), khiến việc tấn công gần như bất khả thi. Nhờ đó, 2FA giúp giảm thiểu đáng kể rủi ro bị xâm nhập trái phép.
 2.  Có thể cải tiến thêm tính năng bảo mật nào cho chương trình này không?
 Có rất nhiều cách để tăng cường bảo mật:

Thêm giới hạn số lần nhập sai mật khẩu, ví dụ sau 3 lần sai sẽ tạm khóa tài khoản.
Thay vì chỉ dùng SHA-256 để băm mật khẩu, nên dùng các thuật toán mạnh hơn như bcrypt hoặc scrypt, có hỗ trợ thêm salt để chống tấn công bảng băm.
Không nên in mã OTP trực tiếp ra màn hình mà nên gửi nó qua email hoặc tin nhắn.
Secret dùng để tạo mã OTP nên được lưu cố định cho từng người dùng, không nên tạo mới mỗi lần chạy chương trình.
Có thể hiển thị mã QR để người dùng quét bằng ứng dụng như Google Authenticator, tạo trải nghiệm thực tế hơn.
Ghi log các lần đăng nhập để phát hiện bất thường.
Mã hóa toàn bộ dữ liệu và sử dụng giao thức HTTPS nếu chạy chương trình trên web.
3. Dựa trên kết quả thực hành, Anh/Chị rút ra được bài học gì về tính bảo mật của mật khẩu và mã OTP?

Dựa trên kết quả thực hành, có thể rút ra một số bài học quan trọng về tính bảo mật của mật khẩu và mã OTP như sau:

 1. Mật khẩu không đủ an toàn nếu chỉ dùng một mình
Khi bạn nhập mật khẩu sai (12345678910), chương trình đã từ chối truy cập — điều này chứng tỏ hệ thống kiểm tra đúng theo cách băm mật khẩu và đối chiếu. Tuy nhiên, trên thực tế:

Mật khẩu rất dễ bị đoán, lộ qua tấn công phishing, hoặc rò rỉ từ cơ sở dữ liệu bị hack.
Nếu không có thêm lớp bảo vệ, người tấn công có thể dễ dàng đăng nhập chỉ với mật khẩu bị đánh cắp.
 2. Mã OTP là lớp bảo mật hiệu quả và khó giả mạo
Mã OTP (ở đây là 161275) chỉ có hiệu lực trong một khoảng thời gian ngắn (thường 30 giây), và được tạo ra dựa trên khóa bí mật và thời gian hiện tại. Trong thực hành:

Sau khi vượt qua bước mật khẩu, bạn phải nhập đúng mã OTP thì mới được xác thực.
Dù ai đó biết được mật khẩu, họ vẫn không thể đăng nhập nếu không có mã OTP hợp lệ tại thời điểm đó.
 3. Sự kết hợp giữa mật khẩu và OTP (2FA) là một giải pháp bảo mật mạnh
Qua thực hành, bạn có thể thấy rõ rằng:

Nếu mật khẩu sai, không thể tiếp tục.
Nếu mật khẩu đúng nhưng mã OTP sai hoặc đã hết hạn, cũng sẽ bị từ chối.
Chỉ khi cả hai đều đúng, người dùng mới được xác thực.
Đây là cơ chế bảo vệ hiệu quả giúp ngăn chặn hầu hết các hình thức truy cập trái phép.

 Bài học rút ra:
Dựa trên thực nghiệm, có thể khẳng định rằng xác thực hai yếu tố (2FA) không chỉ nâng cao bảo mật, mà còn giúp hệ thống trở nên vững chắc trước các cuộc tấn công phổ biến như đánh cắp mật khẩu. Mật khẩu dù mạnh đến đâu cũng nên được kết hợp với OTP hoặc một yếu tố xác thực bổ sung khác để đảm bảo an toàn tối đa cho người dùng.

